rtbkit/common/exchange_connector.h

```diff
@@ -128,6 +128,11 @@ struct ExchangeConnector: public ServiceBase {
     */
     virtual std::string getBidSourceConfiguration() const;
 
+    /** Returns true if bid price is lower then the bidfloor price 
+    */
+
+    virtual bool getBidValidity(const Bid  & bid, const  std::vector<AdSpot> & imp, int & spotIndex) const;
```

Comment:
So `getBidValidity` is now a pretty generic interface. Since users are free to override this function and implement their own logic in their own exchange connector, I would return the "not valid reason" as well, like the `getCampaignCompatibility` function.

You could for example return a `BidValidity` type, which could be a simple struct holding a bool and a std::string, which would be the reason for the bid not to bid valid. You could then use this reason in the `returnInvalidBid` function.